### PR TITLE
perf: Avoid storing DefaultValue in DPDetails

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #

## PR Type

What kind of change does this PR introduce?

## What is the current behavior?

- `DependencyPropertyDetails` object size is unnecessarily large, both directly by storing `_defaultValue` and indirectly through `DependencyPropertyDetails.Stack` size.
- On SamplesApp startup, `DependencyPropertyDetails.Stack` is responsible for 1.27MB of allocations (3.1% of all allocations)
- On SamplesApp startup, `DependencyPropertyDetails` is responsible for 2MB of allocations (4.9% of all allocations)

## What is the new behavior?

- DependencyPropertyDetails size is reduced
- On SamplesApp startup, `DependencyPropertyDetails.Stack` is responsible for 644KB of allocations (1.7%)
- On SamplesApp startup, `DependencyPropertyDetails` is responsible for 1.2MB of allocations (3.3%)

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by MartinZikmund — COMMENTED** [review]
Nice!

**Review by MartinZikmund — APPROVED** [review]

File: build/PackageDiffIgnore.xml [hunk 1]
```diff
@@ -1874,6 +1874,8 @@
 		</Fields>
 
 		<Properties>
+			<Member fullName="Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.FrameworkElement::StretchAffectsMeasureProperty()" reason="Not in WinUI. Made a regular property" />
+			<Member fullName="Microsoft.UI.Xaml.DependencyProperty Microsoft.UI.Xaml.FrameworkElement::StretchAffectsMeasureProperty()" reason="Not in WinUI. Made a regular property" />
```
Comment (jeromelaban) [inline build/PackageDiffIgnore.xml:1878]: Let's keep this change for the next major

File: src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleControl.cs [hunk 2]
```diff
@@ -56,12 +56,16 @@ private static void OnSampleContentChanged(object dependencyObject, DependencyPr
 		}
 
 #if XAMARIN
-		protected override void OnDataContextChanged()
+		internal override bool GetDefaultValue2(DependencyProperty property, out object defaultValue)
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleControl.cs:59]: This should be a breaking change, I'm surprised the validation does not catch it. Let's keep the `OnDataContextChanged` method and mark for next major.

File: src/Uno.UI/UI/Xaml/Documents/TextElement.cs [hunk 3]
```diff
@@ -167,22 +167,27 @@ protected virtual void OnFontSizeChanged()
...
-		private static Brush GetForegroundDefaultValue() => SolidColorBrushHelper.Black;
+		public static DependencyProperty ForegroundProperty { get; } = DependencyProperty.Register(
+			nameof(Foreground),
+			typeof(Brush),
+			typeof(TextElement),
+			new FrameworkPropertyMetadata(
+				defaultValue: SolidColorBrushHelper.Black,
+				options: FrameworkPropertyMetadataOptions.Inherits,
+				propertyChangedCallback: (instance, args) => ((TextElement)instance).OnForegroundChanged()
+		));
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Documents/TextElement.cs:190]: What caused the generator to be unuseable here?

File: build/PackageDiffIgnore.xml (hunk 1 above)
Comment (Youssef1313) [reply build/PackageDiffIgnore.xml:1878]: This property is meant for internal use, and even if someone is using it, they'll be likely using the actual instance property, not this one.

I can still restore though

File: src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleControl.cs (hunk 2 above)
Comment (Youssef1313) [reply src/SamplesApp/SamplesApp.UnitTests.Shared/Controls/UITests/Views/Controls/SampleControl.cs:59]: @jeromelaban This is SamplesApp, so should be okay

File: src/Uno.UI/UI/Xaml/Documents/TextElement.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Documents/TextElement.cs:190]: Due to the way default value work (it's no longer stored, but calculated), we no longer have the backing field update when theme changes. This is also closer to WinUI where property changed isn't fired when theme changes.

**Pull request merged**